### PR TITLE
feat: Define MSRV

### LINKED ISSUE #1310: (feat): clearly define MSRV
# Goal:
Declare a definitive MSRV for FRB.

## Info:
The MSRV of FRB has been debated and mentioned more than a few times, however, we still have nothing better than a guess for what the actual MSRV is.

these are just a few examples:
- https://github.com/fzyzcjy/flutter_rust_bridge/discussions/588
- https://github.com/fzyzcjy/flutter_rust_bridge/pull/1306#discussion_r1279971941
- https://github.com/fzyzcjy/flutter_rust_bridge/pull/1119#issue-1620516613

## Tasks:

- [ ] Find the MSRV (https://github.com/foresterre/cargo-msrv is a wonderful tool)
- [ ] Write toolchain file for cargo (see https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file)
- [ ] Update README.md & docs accordingly
- [ ] CI test MSRV to ensure its integrity (see https://foresterre.github.io/cargo-msrv/ci/index.html)

@fzyzcjy, I'll open a PR in a bit, let me know if you have input.

### PR DESCRIPTION
**Pull request by gutenfries**
# Changes

- Fixes #1310

- Clearly define the MSRV, as well as create necessary files to communicate that to development tools (cargo, clippy, rustup)

- Tests the MSRV in the CI locked version matrix testing

- Adds `cargo-features = ["workspace-inheritance"]` to the head of `Cargo.toml`, which allowed for a lower MSRV without changing any code.

- Adds the MSRV to `README.md`, and creates a page in the docs for the MSRV.

- (UNRELATED) Renamed mismatched `.yml` and `.yaml` files in `.github/workflows/` to all be `.yml`

## Checklist

- [x] add MSRV to `README.md`, as well as the docs
- [x] test MSRV in CI

### CODE REVIEW

**Comment by gutenfries** [comment]
(@fzyzcjy) I've been working on finding an efficient and helpful manner to use the CI to check the MSRV on any Push/PR.

I think we both can see the benefits of this; if the MSRV goes up drastically, perhaps that code should not be merged yet.

In my opinion, a good way to do this is to check the MSRV in GHA, and if the MSRV is _higher_, the action should create a comment on that PR stating the changes, which can then be dealt with.

This differs from a push event, however, and on a push event the MSRV would be checked, and if it has changed, be updated across repo files that reference the MSRV. (this also happens on a PR event)

The issue I am running into is that the `${{ secrets.GITHUB_TOKEN }}` does not have the permissions needed to write comments or create commits (write contents).

(These struggles are the reason for the large amount of commits)

## --

If you think that this is something we should implement, I think that setting up a workflow token for this GHA would be good.

If NOT, then I vote rolling the MSRV verification into `just precommit`, escaping the CI entirely.

Until we agree on how this should be done, and what's best for security, simplicity, etc, I am not going to mark this ready for review, though the POC has been implemented and optimized.

<br/><br/><br/>

Apologies for the quite lengthy summary my friend, and please let me know your thoughts on this

**Comment by fzyzcjy** [comment]
Good job!

Firstly, I guess usually our MSRV should not change (since as a library, we should use the most basic features of Rust to have maximum compability, unless we really need some new feature, or the MSRV is already quite old). Therefore, I expect changing the MSRV is an intentional (explicit) action that the implementor is aware of. For example, a PR may be "let us implement frb feature A which uses Rust language feature B, and since B needs MSRV=1.2.3, our MSRV should be bump to 1.2.3".

Thus, I guess the simplest way is that, we do not check MSRV dynamically. Instead, we run the CI using multiple rust runtime versions (using GHA matrix). For example, say MSRV=1.64, we may test Rust = 1.64 + 1.71.0 (i.e. pinned stable) + 1.73.0 (i.e. pinned nightly). (I am not sure whether we should use something like "nightly" to get latest nightly, or "1.73.0" for pinned nightly, since I do not want to break the CI without any code change which can be annoying).

I can totally understand your feeling if the code you have written has to be changed! So next time maybe we can discuss the rough idea before implementing the details, e.g. if we discuss whether adding MSRV check in CI, I may suggest the ideas above.

So in short, maybe we can remove MSRV check in CI or precommit, but we can add it as a separate command in justfile. Then, we can determine the MSRV (possibly using the script?) and fill that version into CI and readme etc.

**Comment by gutenfries** [comment]
No worries at all my friend!

I had wanted to have an entirely functional POC to provide to you to show the features I had implemented, fully understanding that this is opensource and they may be denied.

with that in mind, all it took was the deletion of one file :)

The dynamic MSRV check may still be run by either `just msrv`, or manually running the scripts in the repo root.

<br/><br/>

My changes are now ready for review

File: .github/workflows/gh-pages.yml
Comment (fzyzcjy) [inline .github/workflows/gh-pages.yml:1]: nit: 👀 these few files are renamed from yaml to yml - if they are done unintentionally unrelated to the PR (e.g. by your IDE automatically) maybe we can restore it to avoid confusion

File: Cargo.toml
Comment (fzyzcjy) [inline Cargo.toml:1]: nit: seems that this is not changed except format, so can restore as well

File: book/src/msrv.md
Comment (fzyzcjy) [inline book/src/msrv.md:1]: nit: we need to add a link to it from our book index

File: book/src/msrv.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+The MSRV (Minimum Supported Rust Version) is the oldest version of Rust that
+this crate is guaranteed to work with.
+
+The MSRV(s) for `flutter_rust_bridge` are as follows:
+
+| Channel | MSRV |
+| ------- | ---- |
+| stable  | <!-- stable msrv --> `1.64+` |
+| nightly | <!-- nightly msrv --> `nightly-2022-07-15+` |
```
Comment (fzyzcjy) [inline book/src/msrv.md:11]: nit: It seems that we are repeating ourself for 3 times: 

1. the one in ci (we test our code with MSRV version)
2. the one in readme.md (the badge)
3. the doc here

is it possible we somehow add a very quick sanity check for them, to ensure that they are always in sync? e.g. some quick python/shell/whatever script to extract MSRV from these 3 places, and check they are equal. Then we can add them to CI, so nobody will accidencially make MSRV not-in-sync

---

EDIT: Oh I see, they are already auto generated from shell. Then what about this:

1. refactor the shell script into "determine MSRV automatically" and "given MSRV, write them into these few places". because in the future someone may want to manually write MSRV.
2. in these few places, maybe we can add comments saying e.g. `<!-- MSRV IS AUTOMATICALLY WRITTEN BY xxxxxxxx.sh, PLEASE DO NOT MODIFY BY HAND, BUT USE THAT SCRIPT-->`

File: tools/stable_msrv.sh
Comment (fzyzcjy) [inline tools/stable_msrv.sh:1]: nit: it seems that stable_msrv.sh and nightly_msrv.sh has a lot of duplicates. in order to have a good maintainability, it would be great to avoid it :)

File: tools/nightly_msrv.sh [hunk 2]
```diff
@@ -0,0 +1,101 @@
+#! /bin/bash
+
+#
+# Ensure all necessary dependencies are installed
+#
+
+if ! command -v rustup >/dev/null 2>&1; then
```
Comment (fzyzcjy) [inline tools/nightly_msrv.sh:7]: nit: will we really meet this situation? IIRC our CI has a step of installing rustup etc. and for local users, if someone does not have rustup, how can he/she develop frb ;)

File: tools/nightly_msrv.sh [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+	# check if the crate is compatible with the version of rust
+	if is_compatible "$mid_date"; then
+		# If compatible, update the end date for the next iteration
+		end_date="$mid_date"
+	else
+		# If not compatible, update the start date for the next iteration
+		start_date=$(date -d "$mid_date + 1 day" +"%Y-%m-%d")
+	fi
+done
```
Comment (fzyzcjy) [inline tools/nightly_msrv.sh:60]: so... it seems that we are repeating ourselves by writing something similar to the https://github.com/foresterre/cargo-msrv you have mentioned? maybe we can call that directly, then we will not need to maintain this script within FRB. (if that package misses some features we need, we can PR to that package first - then, everyone - not only frb developers - will benefit from your code about MSRV!)

File: justfile [hunk 4]
```diff
@@ -194,6 +194,10 @@ ci_codegen:
     just generate_all
     just check_no_git_diff
 
+msrv:
+    bash tools/find_msrv.sh
+    bash tools/find_nightly_msrv.sh
```
Comment (fzyzcjy) [inline justfile:199]: nit: it seems that the files are renamed

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -478,36 +478,45 @@ jobs:
...
 
-  # -------------- MSRV & specific version testing --------------
   msrv:
-    name: MSRV & specific version testing
+    name: test MSRV and specific problematic versions
     runs-on: ubuntu-latest # as of now, there is no real reason to test on multiple OSs
     strategy:
       matrix:
        rust_version:
+         - nightly              # latest nightly
```
Comment (fzyzcjy) [inline .github/workflows/ci.yml:487]: nit: also add latest stable

File: clippy.toml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+# https://doc.rust-lang.org/nightly/clippy/lint_configuration.html#avoid-breaking-exported-api
+avoid-breaking-exported-api = true
+
+# https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_methods
+disallowed-methods = [
+	"std::time::Instant::now",      # use `instant` crate instead for wasm/web compatibility
+	"std::time::Duration::elapsed", # use `instant` crate instead for wasm/web compatibility
+	"std::time::SystemTime::now",   # use `instant` or `time` crates instead for wasm/web compatibility
+	"std::thread::spawn",           # Cannot spawn threads on wasm
+	# TODO(@fzyzcjy): add methods here as needed
```
Comment (fzyzcjy) [inline clippy.toml:12]: no I do not have suggestions here (yet), so the todo can be removed ;)

File: clippy.toml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+]
+
+# https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_types
+disallowed-types = [
+	# TODO(@fzyzcjy): add types here as needed
```
Comment (fzyzcjy) [inline clippy.toml:18]: same

File: frb_codegen/Cargo.toml [hunk 8]
```diff
@@ -1,10 +1,12 @@
+cargo-features = ["workspace-inheritance"]
```
Comment (fzyzcjy) [inline frb_codegen/Cargo.toml:1]: nit: so... why do we need this feature? (guess it is because clippy?)

File: frb_rust/Cargo.toml [hunk 9]
```diff
@@ -31,15 +33,15 @@ console_error_panic_hook = { version = "0.1.7", optional = true }
...
-  "DedicatedWorkerGlobalScope",
-  "MessageEvent",
-  "MessagePort",
-  "ErrorEvent",
-  "Blob",
-  "BlobPropertyBag",
-  "Worker",
-  "Url",
-  "BroadcastChannel",
+	"DedicatedWorkerGlobalScope",
```
Comment (fzyzcjy) [inline frb_rust/Cargo.toml:36]: nit: it would be great if the old format could be kept, because someone else in the future may also automatically change the format via IDE automatically, then the code blame history will be a mess... but that's not a super big problem.

**Review by fzyzcjy — COMMENTED** [review]
Good job! Only some nits (though look like a dozen, but they are mostly quite trivial)

File: .github/workflows/gh-pages.yml
Comment (gutenfries) [reply .github/workflows/gh-pages.yml:1]: I intentionally renamed them.

If you prefer `.yaml`, that's okay as well, I was just confused by the inconsistent pattern.

If there is a reason that some of the `.github/workflows/` files are `.yaml` and others `.yml`, I would be curious to understand why.

We can cherry-pick past this and address it in another PR (or push)

File: book/src/msrv.md (hunk 1 above)
Comment (gutenfries) [reply book/src/msrv.md:11]: I had intended to do this, ill get it fixed :)

File: frb_codegen/Cargo.toml (hunk 8 above)
Comment (gutenfries) [reply frb_codegen/Cargo.toml:1]: older versions of cargo require this line to be present in order to use workspace dependency inheritance. (however, is now unneeded)

By adding this line alone, the MSRV dropped quite a bit.

File: frb_rust/Cargo.toml (hunk 9 above)
Comment (gutenfries) [reply frb_rust/Cargo.toml:36]: what toml formatter does the project use?

If it is [Taplo](https://taplo.tamasfe.dev), perhaps a configuration file may be helpful (https://taplo.tamasfe.dev/configuration/file.html)

File: Cargo.toml
Comment (gutenfries) [reply Cargo.toml:1]: resolving as duplicate

File: tools/nightly_msrv.sh (hunk 2 above)
Comment (gutenfries) [reply tools/nightly_msrv.sh:7]: You are correct in your thinking, HOWEVER, remember this is bash. If a command isn't found, the script continues anyway, potentially causing rather large issues.

three lines can make CERTAIN rustup is available on the system with a clear mind.

Perhaps someone is using a new shell that does not have `.rustup/` in the shell-loaded path?

It is an edge case certainly, but it is common good practice to ensure all tools are installed at the beginning of bash scripts for this reason.

File: tools/nightly_msrv.sh (hunk 3 above)
Comment (gutenfries) [reply tools/nightly_msrv.sh:60]: The reasons I decided to implement this from scratch are:

a) https://github.com/foresterre/cargo-msrv does NOT support nightly (perhaps I'll make a PR there in the future) and

b) implementing it from scratch allows for find-and-replace to automate updating/adding versions to CI, README, book, etc.

File: .github/workflows/gh-pages.yml
Comment (fzyzcjy) [reply .github/workflows/gh-pages.yml:1]: Ah I did not notice that (only checked the PR file panel), then yes we should make it consistent!

File: frb_rust/Cargo.toml (hunk 9 above)
Comment (fzyzcjy) [reply frb_rust/Cargo.toml:36]: You are right, there is no format yet. I am just following flutter's convention: when there is no (auto) formatter, try to not change the code due to format.

However, feel free to configure a formatter (e.g. Taplo)! Just ensure that

- [ ] it is checked in CI (along with other formatters, like Dart and Rust formatter)
- [ ] it is run automatically when `just precommit` (trivial to do if you put is as a sibling to the existing Dart format etc)
- [ ] it has nice support and is popular, e.g. has VSCode and Intellij IDEA bulit-in support or plugin support

File: tools/nightly_msrv.sh (hunk 2 above)
Comment (fzyzcjy) [reply tools/nightly_msrv.sh:7]: > HOWEVER, remember this is bash. If a command isn't found, the script continues anyway, potentially causing rather large issues.

I have learnt to set `set -euxo pipefail` *whenever* you are writing any bash script. Then if a command is not found (or bad return code, or whatever), you are guaranteed to fail immediately.

File: tools/nightly_msrv.sh (hunk 3 above)
Comment (fzyzcjy) [reply tools/nightly_msrv.sh:60]: > a) [foresterre/cargo-msrv](https://github.com/foresterre/cargo-msrv) does NOT support nightly (perhaps I'll make a PR there in the future)

Yes I think so, that's why I hope you to PR there - then everyone in Rust ecosystem will benefit from your code about MSRV.

> b) implementing it from scratch allows for find-and-replace to automate updating/adding versions to CI, README, book, etc.

To have a good separate of concerns, it would be great to extract it to a function separate from determining MSRV (as is mentioned in another comment)

Comment (gutenfries) [reply tools/nightly_msrv.sh:60]: > > a) [foresterre/cargo-msrv](https://github.com/foresterre/cargo-msrv) does NOT support nightly (perhaps I'll make a PR there in the future)
> 
> Yes I think so, that's why I hope you to PR there - then everyone in Rust ecosystem will benefit from your code about MSRV.

I whole-heartedly agree.

To be completely honest, I think that this PR should be closed, and once I/the dev team @ https://github.com/foresterre/cargo-msrv implement and release nightly MSRV checking, another different PR should be reviewed.

there are too many irrelevant commits in this PR and simply put, there is no reason.

I'll probably open a draft PR to track progress, but I'll mark as ready for merge and send a ping when we're ready.

Apologies my friend for the lack of communication prior to opening this PR

**Comment by gutenfries** [comment]
> > a) [foresterre/cargo-msrv](https://github.com/foresterre/cargo-msrv) does NOT support nightly (perhaps I'll make a PR there in the future)
> 
> Yes I think so, that's why I hope you to PR there - then everyone in Rust ecosystem will benefit from your code about MSRV.

I whole-heartedly agree.

To be completely honest, I think that this PR should be closed, and once I/the dev team @ https://github.com/foresterre/cargo-msrv implement and release nightly MSRV checking, another different PR should be reviewed.

there are too many irrelevant commits in this PR and simply put, there is no reason.

I'll probably open a draft PR to track progress, but I'll mark as ready for merge and send a ping when we're ready.

**UPDATE:** (https://github.com/foresterre/cargo-msrv/issues/101#issuecomment-1665853060) this is actively being worked on.

**Comment by fzyzcjy** [comment]
> To be completely honest, I think that this PR should be closed, and once I/the dev team @ [foresterre/cargo-msrv](https://github.com/foresterre/cargo-msrv) implement and release nightly MSRV checking, another different PR should be reviewed.
> there are too many irrelevant commits in this PR and simply put, there is no reason.

I agree. Borrowing Flutter's PR philosophy - one PR should only do one thing, it looks good to split this PR into multiple PRs.

> I'll probably open a draft PR to track progress, but I'll mark as ready for merge and send a ping when we're ready.

Sure, take your time!

> UPDATE: (https://github.com/foresterre/cargo-msrv/issues/101#issuecomment-1665853060) this is actively being worked on.

That looks great :)
